polygeist/tools/cgeist/Lib/CGCall.cc

```diff
@@ -79,12 +80,82 @@ static void castCallerArgs(mlir::func::FuncOp callee,
   }
 }
 
+/// Return whether a builder is or not inserting in a device function.
+static bool inDeviceFunc(OpBuilder &Builder) {
+  auto Func =
+      Builder.getInsertionBlock()->getParent()->getParentOfType<func::FuncOp>();
+  return !Func || mlirclang::isDeviceFunction(Func);
```

Comment:
If `Func` is not found this function assumes the builder injects into a device function, I think this should be:

     `return Func && mlirclang::isDeviceFunction(Func);`

Reply:
In fact, if `Func` is not found, it means we are inserting in a `GPUFuncOp`, so we must be inserting in the device module.